examples/pytorch/GATNE-T/src/utils.py

```diff
@@ -0,0 +1,191 @@
+                        help='Negative samples for optimization. Default is 5.')
+    
+    parser.add_argument('--neighbor-samples', type=int, default=10,
+                        help='Neighbor samples for aggregation. Default is 10.')
+
+    parser.add_argument('--patience', type=int, default=5,
+                        help='Early stopping patience. Default is 5.')
+    
+    return parser.parse_args()
+
+
+# for each line, the data is [edge_type, node, node]
+def load_training_data(f_name):
+    print('We are loading data from:', f_name)
+    edge_data_by_type = dict()
+    all_nodes = list()
+    with open(f_name, 'r') as f:
+        for line in f:
+            words = line[:-1].split(' ')  # line[-1] == '\n'
+            if words[0] not in edge_data_by_type:
+                edge_data_by_type[words[0]] = list()
+            x, y = words[1], words[2]
+            edge_data_by_type[words[0]].append((x, y))
+            all_nodes.append(x)
+            all_nodes.append(y)
+    all_nodes = list(set(all_nodes))
+    print('Total training nodes: ' + str(len(all_nodes)))
+    return edge_data_by_type
+
+
+# for each line, the data is [edge_type, node, node, true_or_false]
+def load_testing_data(f_name):
+    print('We are loading data from:', f_name)
+    true_edge_data_by_type = dict()
+    false_edge_data_by_type = dict()
+    all_edges = list()
+    all_nodes = list()
+    with open(f_name, 'r') as f:
+        for line in f:
+            words = line[:-1].split(' ')
+            x, y = words[1], words[2]
+            if int(words[3]) == 1:
+                if words[0] not in true_edge_data_by_type:
+                    true_edge_data_by_type[words[0]] = list()
+                true_edge_data_by_type[words[0]].append((x, y))
+            else:
+                if words[0] not in false_edge_data_by_type:
+                    false_edge_data_by_type[words[0]] = list()
+                false_edge_data_by_type[words[0]].append((x, y))
+            all_nodes.append(x)
+            all_nodes.append(y)
+    all_nodes = list(set(all_nodes))
+    return true_edge_data_by_type, false_edge_data_by_type
+
+
+def load_node_type(f_name):
+    print('We are loading node type from:', f_name)
+    node_type = {}
+    with open(f_name, 'r') as f:
+        for line in f:
+            items = line.strip().split()
+            node_type[items[0]] = items[1]
+    return node_type
+
+
+def generate_pairs(all_walks, window_size):
+    # for each node, choose the first neighbor and second neighbor of it to form pairs
+    pairs = []
+    skip_window = window_size // 2
+    for layer_id, walks in enumerate(all_walks):
+        for walk in walks:
+            for i in range(len(walk)):
+                for j in range(1, skip_window + 1):
+                    if i - j >= 0:
+                        pairs.append((walk[i], walk[i-j], layer_id))
+                    if i + j < len(walk):
+                        pairs.append((walk[i], walk[i+j], layer_id))
```

Comment:
this looks very slow.

Reply:
This can be regarded as part of data preprocessing, and it does not affect training speed that much.